src/Http/Routing/src/DefaultLinkGenerator.cs

```diff
@@ -336,137 +336,92 @@ public void Dispose()
                 }
             }
 
+            [LoggerMessage(EventId = 103, EventName = "TemplateFailedRequiredValues", Level = LogLevel.Debug,
+                Message = "Failed to process the template {Template} for {Endpoint}. " +
+                "A required route value is missing, or has a different value from the required default values. " +
+                "Supplied ambient values {AmbientValues} and {Values} with default values {Defaults}")]
+            public static partial void TemplateFailedRequiredValues(ILogger logger, string template, string endpoint, string ambientValues, string values, string defaults);
+
             public static void TemplateFailedConstraint(ILogger logger, RouteEndpoint endpoint, string parameterName, IRouteConstraint constraint, RouteValueDictionary values)
             {
                 // Checking level again to avoid allocation on the common path
                 if (logger.IsEnabled(LogLevel.Debug))
                 {
-                    _templateFailedConstraint(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, constraint, parameterName, FormatRouteValues(values), null);
+                    TemplateFailedConstraint(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, parameterName, constraint, FormatRouteValues(values));
                 }
             }
 
+            [LoggerMessage(EventId = 103, EventName = "TemplateFailedConstraint", Level = LogLevel.Debug,
+                Message = "Failed to process the template {Template} for {Endpoint}. " +
+                "The constraint {Constraint} for parameter {ParameterName} failed with values {Values}")]
+            public static partial void TemplateFailedConstraint(ILogger logger, string template, string endpoint, string parameterName, IRouteConstraint constraint, string values);
+
             public static void TemplateFailedExpansion(ILogger logger, RouteEndpoint endpoint, RouteValueDictionary values)
             {
                 // Checking level again to avoid allocation on the common path
                 if (logger.IsEnabled(LogLevel.Debug))
                 {
-                    _templateFailedExpansion(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, FormatRouteValues(values), null);
+                    TemplateFailedExpansion(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, FormatRouteValues(values));
                 }
             }
 
+            [LoggerMessage(EventId = 104, EventName = "TemplateFailedExpansion", Level = LogLevel.Debug,
+                Message = "Failed to process the template {Template} for {Endpoint}. " +
+                "The failure occurred while expanding the template with values {Values} " +
+                "This is usually due to a missing or empty value in a complex segment")]
+            public static partial void TemplateFailedExpansion(ILogger logger, string template, string endpoint, string values);
```

Comment:
Is skip log level check added yet? Can add it to attribute when method is called from an overload that already checks for IsEnabled.

Reply:
Not yet but close: https://github.com/dotnet/runtime/pull/54305

Add comments to attributes where it should be added once supported.